client_internals_it_test.go

```diff
@@ -542,3 +543,176 @@ func ensureClient(config hz.Config) *hz.Client {
+			wantErr:  false,
+		},
+		{
+			name:     "testMA_OptionalServerNotAuthenticated",
+			muStatus: OPTIONAL,
+			ca:       "testdata/unknown-server-cert.pem",
+			cert:     "testdata/OpenSSL/client/client.crt",
+			key:      "testdata/OpenSSL/client/client.key",
+			wantErr:  true,
+		},
+		{
+			name:     "testMA_OptionalClientNotAuthenticated",
+			muStatus: OPTIONAL,
+			ca:       "testdata/OpenSSL/rootCA.crt",
+			cert:     "testdata/unknown-client-cert.pem",
+			key:      "testdata/unknown-client-key.pem",
+			wantErr:  true,
+		},
+		{
+			name:     "testMA_OptionalClientAndServerNotAuthenticated",
+			muStatus: OPTIONAL,
+			ca:       "testdata/unknown-server-cert.pem",
+			cert:     "testdata/unknown-client-cert.pem",
+			key:      "testdata/unknown-client-key.pem",
+			wantErr:  true,
+		},
+		{
+			name:     "testMA_RequiredWithNoCertFile",
+			muStatus: REQUIRED,
+			ca:       "testdata/OpenSSL/rootCA.crt",
+			cert:     "",
+			key:      "",
+			wantErr:  true,
+		},
+		{
+			name:     "testMA_OptionalWithNoCertFile",
+			muStatus: OPTIONAL,
+			ca:       "testdata/OpenSSL/rootCA.crt",
+			cert:     "",
+			key:      "",
+			wantErr:  false,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			skip.IfNot(t, "oss")
+			if tc.wantErr {
+				require.Error(t, sslTestUtil(t, tc.muStatus, tc.ca, tc.cert, tc.key))
+			} else {
+				require.NoError(t, sslTestUtil(t, tc.muStatus, tc.ca, tc.cert, tc.key))
+			}
+		})
+	}
+}
+
+func sslTestUtil(t *testing.T, mu MutualAuthenticationStatus, ca, cert, key string) error {
+	ctx := context.Background()
+	port := it.NextPort()
+	clsConfig := xmlSSLMutualAuthenticationConfig(t.Name(), port, mu)
+	tc := it.StartNewClusterWithConfig(1, clsConfig, port)
+	defer tc.Shutdown()
+	clientCfg := tc.DefaultConfig()
+	if err := clientCfg.Cluster.Network.SSL.SetCAPath(ca); err != nil {
+		return err
+	}
+	// unset certificate and key should not be added
+	if len(cert) != 0 && len(key) != 0 {
+		if err := clientCfg.Cluster.Network.SSL.AddClientCertAndKeyPath(cert, key); err != nil {
+			return err
+		}
+	}
+	client, err := hz.StartNewClientWithConfig(ctx, clientCfg)
+	if err != nil {
+		return err
+	}
+	if err = client.Shutdown(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func xmlSSLMutualAuthenticationConfig(clusterName string, port int, mu MutualAuthenticationStatus) string {
+	return fmt.Sprintf(`
+		<hazelcast xmlns="http://www.hazelcast.com/schema/config"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://www.hazelcast.com/schema/config
+           http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd">
+			<cluster-name>%s</cluster-name>
+			<network>
+				<port>%d</port>
+				<ssl enabled="true">
+					<factory-class-name>
+						com.hazelcast.nio.ssl.ClasspathSSLContextFactory
+					</factory-class-name>
+					<properties>
+						<property name="keyStore">testdata/OpenSSL/ca.keystore</property>
```

Comment:
This assumes RC is being run in the project root:
```
<property name="keyStore">testdata/OpenSSL/ca.keystore</property>
```
I think that should be more flexible, and use an absolute path instead of the relative path. The absolute path can be found out using something like:
```
wd, err := os.Getwd()
path := filepath.Join(wd, "testdata/OpenSSL/ca.keystore")
```

Reply:
Absolute path could not be resolved for the certificate artifacts, the only working way is putting resources to the root directory, then using relative path. I could not figure out the reason.